### DISCUSSION TITLE
Backend Validation on Schema forms

### DISCUSSION THREAD

**Discussion by CoolGoose** [formkit/formkit#1452] [General]
Hey folks,

Does anyone have a best tip / example on how you would also do inline per field validation when you wan to make sure the backend also agrees with the frontend ?

Is a generic backend validator a bad idea ? Or is something similar already built and I can't find it ? :D

**Comment by pjrusso** [comment]
I had an idea to use Zod to define the validation rules (https://zod.de) rather than the native FormKit validators and then on the backend (ASP.NET Core in my case), populate that schema with the POSTed data and re-use the same Zod rules using the JINT library (see https://github.com/sebastienros/jint).

I had some success in a previous role executing the Javacript rules and calculations for a bespoke forms solutions using JINT (JINT isnt' a wrapper around a Javascript engine as such, so can be considered much lighter weight).

I need my FormKit forms to be dynamically generated - so hence not being able to get to implementing the above yet though - sorry!

**Comment by pjrusso** [comment]
I have a quick test case [here](https://formkit.link/fd16da01222fe49d86a10f4fe9752814/) but I'm not happy with the way in which I have to do the Zod 'conditional' rules - but I'll drop it into ASP.NET Core Jint and see what happens?

**Reply by justin-schroeder** [reply]
Did you see there is a first-party zod library? https://formkit.com/plugins/zod

**Reply by pjrusso** [reply]
Thanks @justin-schroeder - I did yes - I was new to Zod back in January I realise my reply above was a bit sparse.

I have played around with the Zod plugin stuff only recently but what I came up against was the difficulty in defining conditional rules in Zod (to me) - which isn't a show stopper but it needs some further thought from me as to how to generate this from my abstracted form definition so I can use the Zod rules in my Jint-enabled server back-stop.

One issue I did find today is that using Zod doesn't play with the '$state.required' and '$state.invalid' mechanism at present - but again not a showstopper as that was only to drive the '*' for mandatory fields and to generate the 'aria-invalid' attributes so far.

To be honest though - and hats off to you guys - most of the other forms-related I've looked at over the years either required a huge investment in bespoke solutions front and back or if choosing an off-the-shelf solution (sold to a business with a fancy looking form builder...) contained a "_you are ultimately responsible for validating user-form submissions server-side yourself_" in the small print - but you couldn't easily share the business logic front and back (so I suspect that there is lot out there that doesn't bother) - but you are addressing that right from the off.

As I am generating my FormKit form schema dynamically - then I am comfortable that my solution may be to generate the dynamic 'if' and 'if-then-else' validators myself and then auto generate something I can run server-side to join the two together.
